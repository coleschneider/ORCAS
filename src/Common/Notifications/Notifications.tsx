import * as React from 'react';
import Idle from 'react-idle';
import { NETWORK_MESSAGES } from 'utils/messages';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';



export default class Notifications extends React.Component {
  idleToast = null;
  networkToast = null;
  idleTimer;
  constructor(props) {
    super(props);
  }
  componentDidMount() {
    window.addEventListener('online', this.handleNetworkChange);
    window.addEventListener('offline', this.handleNetworkChange);
  }
  handleNetworkChange = ({ type }) => {
    const { style, text } = NETWORK_MESSAGES[type];
    this.networkToast = toast[style](text, {
      autoClose: 5000,
    });
  };

  onIdle = ({ idle }) => {
    if (this.idleToast === null && idle) {
      this.idleToast = toast.warn(
        `Looks like you've been inactive for a while. We make sure to clear personal information after an hour of inactivity.`,
        {
          autoClose: false,
          position: 'bottom-right',
        },
      );
    } else {
      this.onActive()
    }
  }
  onActive = () => {
    toast.update(this.idleToast, {
      autoClose: 3000,
      onClose: () => {
        this.idleToast = null;
      },
    });
  }
  render() {
    return (
      <div>
        <Idle 
          onChange={this.onIdle}
          timeout={1000 * 60 * 5} /* 5 minutes */
          />
        <ToastContainer />
      </div>
    );
  }
}
