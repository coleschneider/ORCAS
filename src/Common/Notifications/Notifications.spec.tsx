import * as React from 'react';
import Notifications from './Notifications';
import { toast } from 'react-toastify';

jest.useFakeTimers();

describe('Notifications', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<Notifications />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should call the timer', () => {
    const wrapper = mount(<Notifications />);
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 0);
  });
  it('should create a toast when onIdle is called', () => {
    const wrapper = mount(<Notifications />);
    wrapper.instance().onIdle({ idle: true });
    wrapper.update();
    expect(wrapper.instance().idleToast).toBeDefined();
  });
  it('should update the toast when the toast is defined and its not in idle', () => {
    const wrapper = mount(<Notifications />);
    const toastSpy = jest.spyOn(toast, 'update');

    const idleToastId = 'asdad';
    wrapper.instance().idleToast = idleToastId;

    wrapper.instance().onIdle({ idle: false });

    expect(toastSpy).toHaveBeenCalledWith(idleToastId, { autoClose: 3000, onClose: expect.any(Function) });
  });
  it('should set the idleToast instance to null after seconds following activity', () => {
    const wrapper = mount(<Notifications />);
    const onCloseSpy = jest.spyOn(toast, 'update');
    wrapper.instance().onIdle({ idle: true });
    wrapper.instance().onActive();
    setTimeout(() => {
      expect(wrapper.instance().idleToast).toBe(null);
    }, 5000);

    expect(onCloseSpy).toBeCalledWith(
      expect.any(String),
      expect.objectContaining({
        onClose: expect.any(Function),
      }),
    );
  });
});
