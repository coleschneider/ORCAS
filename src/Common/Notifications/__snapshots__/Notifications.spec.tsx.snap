// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications should match the snapshot 1`] = `
<div>
  <Idle
    defaultIdle={false}
    events={
      Array [
        "mousemove",
        "mousedown",
        "keydown",
        "touchstart",
        "scroll",
      ]
    }
    onChange={[Function]}
    render={[Function]}
    timeout={300000}
  />
  <ToastContainer
    autoClose={5000}
    bodyClassName={null}
    className={null}
    closeButton={
      <CloseButton
        ariaLabel="close"
      />
    }
    closeOnClick={true}
    draggable={true}
    draggablePercent={80}
    hideProgressBar={false}
    newestOnTop={false}
    pauseOnFocusLoss={true}
    pauseOnHover={true}
    position="top-right"
    progressClassName={null}
    progressStyle={null}
    rtl={false}
    style={null}
    toastClassName={null}
    transition={[Function]}
  />
</div>
`;
